### DISCUSSION TITLE
ScatterplotLayer custom getPosition attribute

### DISCUSSION THREAD

**Discussion by esiegel** [visgl/deck.gl#5496] [Q&A] [answered]
What is the correct way of setting a custom `getPosition` attribute for the ScatterplotLayer?

I have an array `positions` that is a typed `Float64Array` with data like [x,y,z,...] and have passed it as such:

```javascript
  const layer = new ScatterplotLayer({
    filled: true,
    getFillColor: [100, 130, 150],
    getLineColor: [0, 0, 0],
    id: "scatterplot-layer",
    data: {
      attributes: {
        getPosition: {
          size: 3,
          value: positions,
        },
      },
    },
  });
```

This ends up rendering a single filled point and I'm not sure why.  There should be around 1.3 million.  If instead of using a custom attribute I use a standard `data: d => some_logic` it works just fine.  I wonder if this has something to do with `fp64` and internal logic on when the float64Low buffer needs to be created?

Other things I have tried:

- Using a `Float32Array`
- Fiddling with `stride` `type`

The last point of info, comes from `luma.gl` debugging logs that shows that in the good path I am passing twice as much data to the gpu.

```
// GOOD CASE using data => some_logic
3: instancePositions (vec3)	"*[8.5,-0.47,0., 4.6e-7,1.4e-8,0., 3.6,0.17,0., -2.0e-8]"	"I 5224512 (x3=62694144 bytes GL.FLOAT)"
5: instancePositions64Low (vec3)	"[8.5,-0.47,0., 4.6e-7,1.4e-8,0., 3.6,0.17,0., -2.0e-8]"	"I 5224512 (x3=62694144 bytes GL.FLOAT)"


// BAD CASE using custom attribute
3: instancePositions (vec3)	"*[8.5,-0.47,0., 4.6e-7,1.4e-8,0., 3.6,0.17,0., -2.0e-8]"	"I 2612258 (x3=31347096 bytes GL.FLOAT)"
5: instancePositions64Low (vec3)	"[8.5,-0.47,0., 4.6e-7,1.4e-8,0., 3.6,0.17,0., -2.0e-8]"	"I 2612258 (x3=31347096 bytes GL.FLOAT)"
```

Thanks for any and all help.

**Comment by Pessimistress** [comment, marked as answer]
If you use a non-iterable `data` object it must contain a `length` field: https://deck.gl/docs/api-reference/core/layer#data

In your case you need to provide

```js
{
  length: POINT_COUNT,
  attributes: {...}
}
```

**Reply by esiegel** [reply]
And I thought I was reading the docs carefully.  Apologies for the false alarm, and thank you for the help.
